File: src/game/trainingPipeline.ts
/**
 * @file trainingPipeline.ts
 * @module game/trainingPipeline
 * @layer Layer 12, Layer 13, Layer 14
 * @component Training Pipeline — Game Events → SFT → HuggingFace Fine-Tune
 *
 * How it works:
 *
 *   1. Player PAYS Energy Tokens for an adventure (dungeon run, tower floor, etc.)
 *   2. During the adventure, game events stream into a 3-tier dataset:
 *        RAW → QUARANTINED → APPROVED
 *   3. Approved events are formatted as SFT (Supervised Fine-Tune) pairs:
 *        { instruction: <game_context>, response: <companion_action> }
 *   4. SFT pairs are batched and pushed to HuggingFace dataset repo.
 *   5. When enough data accumulates, a fine-tune job trains the companion model.
 *   6. The updated model weights deploy back into the companion's AI.
 *
 * This means: every adventure the player pays for TRAINS their companion.
 * The companion literally learns from the player's gameplay decisions.
 *
 * A3: Causality — events are time-ordered, append-only.
 * A5: Composition — full audit trail from payment → events → SFT → model.
 */

import { ConsumptionRecord } from './energyToken.js';

// ---------------------------------------------------------------------------
//  Event Types — What happens during an adventure
// ---------------------------------------------------------------------------

/** Categories of game events that generate training data */
export type GameEventType =
  | 'combat_action'       // Player chose an attack / defense / dodge
  | 'companion_command'   // Player issued a companion command
  | 'companion_response'  // Companion autonomous behavior
  | 'evolution_choice'    // Player selected an evolution branch
  | 'formation_change'    // Player rearranged fleet formation
  | 'codex_query'         // Player used codex terminal
  | 'npc_dialogue'        // Player made a dialogue choice
  | 'item_use'            // Player used an item
  | 'exploration_action'  // Player explored / solved puzzle
  | 'tower_strategy';     // Tower floor tactical decision

/** A single game event captured during an adventure */
export interface GameEvent {
  readonly eventId: string;
  readonly sessionId: string;
  readonly playerId: string;
  readonly companionId: string;
  readonly eventType: GameEventType;
  readonly timestamp: number;

  /** The game state context BEFORE the event */
  readonly context: EventContext;

  /** The action that was taken */
  readonly action: EventAction;

  /** The outcome that resulted */
  readonly outcome: EventOutcome;
}

/** Snapshot of game state at the moment of the event */
export interface EventContext {
  /** Current tongue vector (6D) */
  readonly tongueVector: readonly number[];
  /** Current floor / area */
  readonly location: string;
  /** Companion's current HP ratio (0-1) */
  readonly companionHpRatio: number;
  /** Number of enemies present */
  readonly enemyCount: number;
  /** Current formation type */
  readonly formationType: string;
  /** Bond level with companion */
  readonly bondLevel: number;
  /** Companion evolution stage */
  readonly evolutionStage: string;
  /** Any active status effects */
  readonly statusEffects: readonly string[];
}

/** What action was taken */
export interface EventAction {
  /** Action identifier */
  readonly actionId: string;
  /** Human-readable description */
  readonly description: string;
  /** Whether this was a player choice or companion autonomous */
  readonly source: 'player' | 'companion' | 'system';
  /** Confidence of companion action (0-1, only for companion source) */
  readonly confidence: number;
}

/** What happened as a result */
export interface EventOutcome {
  /** Was it successful? */
  readonly success: boolean;
  /** Damage dealt / healed / etc */
  readonly numericResult: number;
  /** Text description */
  readonly description: string;
  /** Any tongue vector changes */
  readonly tongueShift: readonly number[] | null;
  /** SCBE safety score for this event */
  readonly safetyScore: number;
}

// ---------------------------------------------------------------------------
//  3-Tier Dataset — RAW → QUARANTINED → APPROVED
// ---------------------------------------------------------------------------

/** Dataset tiers — events must be promoted through each tier */
export type DatasetTier = 'RAW' | 'QUARANTINED' | 'APPROVED';

/** A dataset record wrapping a game event with tier metadata */
export interface DatasetRecord {
  readonly recordId: string;
  readonly event: GameEvent;
  tier: DatasetTier;
  readonly createdAt: number;
  promotedAt: number | null;
  /** Why it was quarantined (if applicable) */
  quarantineReason: string | null;
  /** Review notes */
  reviewNotes: string | null;
}

// ---------------------------------------------------------------------------
//  SFT Pair — The actual training data format
// ---------------------------------------------------------------------------

/** An SFT (Supervised Fine-Tuning) pair derived from approved events */
export interface SFTPair {
  readonly pairId: string;
  readonly sourceRecordId: string;
  readonly companionId: string;

  /** The instruction / prompt for the model */
  readonly instruction: string;

  /** The expected response (what the companion should learn to do) */
  readonly response: string;

  /** Category for dataset organization */
  readonly category: string;

  /** Quality score (0-1) based on outcome success and safety */
  readonly qualityScore: number;

  readonly timestamp: number;
}

/** HuggingFace push batch */
export interface HFBatch {
  readonly batchId: string;
  readonly companionId: string;
  readonly pairs: readonly SFTPair[];
  readonly datasetRepo: string;
  readonly pushedAt: number | null;
  readonly status: 'pending' | 'pushed' | 'failed';
}

/** Fine-tune job tracking */
export interface FineTuneJob {
  readonly jobId: string;
  readonly companionId: string;
  readonly datasetRepo: string;
  readonly modelRepo: string;
  readonly batchIds: readonly string[];
  readonly totalPairs: number;
  readonly startedAt: number;
  completedAt: number | null;
  status: 'queued' | 'running' | 'completed' | 'failed';
  /** Model checkpoint hash when complete */
  checkpointHash: string | null;
}

// ---------------------------------------------------------------------------
//  Safety Thresholds
// ---------------------------------------------------------------------------

/** Minimum SCBE safety score to auto-approve an event */
export const AUTO_APPROVE_THRESHOLD = 0.8;

/** Below this score, event goes to QUARANTINED instead of APPROVED */
export const QUARANTINE_THRESHOLD = 0.4;

/** Minimum quality score for an SFT pair to be included in training */
export const MIN_QUALITY_SCORE = 0.5;

/** Number of approved pairs needed before triggering a fine-tune */
export const FINE_TUNE_THRESHOLD = 100;

// ---------------------------------------------------------------------------
//  Training Pipeline
// ---------------------------------------------------------------------------

export class TrainingPipeline {
  private _companionId: string;
  private _playerId: string;
  private _records: DatasetRecord[] = [];
  private _pairs: SFTPair[] = [];
  private _batches: HFBatch[] = [];
  private _jobs: FineTuneJob[] = [];
  private _datasetRepo: string;
  private _modelRepo: string;

  constructor(
    companionId: string,
    playerId: string,
    datasetRepo: string = '',
    modelRepo: string = ''
  ) {
    this._companionId = companionId;
    this._playerId = playerId;
    this._datasetRepo = datasetRepo || `scbe-companion-${companionId}`;
    this._modelRepo = modelRepo || `scbe-model-${companionId}`;
  }

  get companionId(): string {
    return this._companionId;
  }

  get datasetRepo(): string {
    return this._datasetRepo;
  }

  get modelRepo(): string {
    return this._modelRepo;
  }

  // -------------------------------------------------------------------------
  //  Tier 1: Ingest — RAW events
  // -------------------------------------------------------------------------

  /**
   * Ingest a game event into the RAW tier.
   * Returns the dataset record ID.
   */
  ingestEvent(event: GameEvent): string {
    const recordId = `rec_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;

    const record: DatasetRecord = {
      recordId,
      event,
      tier: 'RAW',
      createdAt: Date.now() / 1000,
      promotedAt: null,
      quarantineReason: null,
      reviewNotes: null,
    };

    this._records.push(record);
    return recordId;
  }

  /**
   * Bulk-ingest events from an adventure session.
   * Automatically classifies into QUARANTINED or APPROVED based on safety score.
   */
  ingestSession(events: GameEvent[]): { approved: number; quarantined: number; rejected: number } {
    let approved = 0;
    let quarantined = 0;
    let rejected = 0;

    for (const event of events) {
      const recordId = this.ingestEvent(event);
      const result = this.classifyRecord(recordId);

      if (result === 'APPROVED') approved++;
      else if (result === 'QUARANTINED') quarantined++;
      else rejected++;
    }

    return { approved, quarantined, rejected };
  }

  // -------------------------------------------------------------------------
  //  Tier 2: Classify — RAW → QUARANTINED or APPROVED
  // -------------------------------------------------------------------------

  /**
   * Classify a RAW record based on its safety score.
   * High safety = auto-approve. Low safety = quarantine. Very low = reject.
   */
  classifyRecord(recordId: string): DatasetTier | 'REJECTED' {
    const record = this._records.find((r) => r.recordId === recordId);
    if (!record || record.tier !== 'RAW') return 'REJECTED';

    const safetyScore = record.event.outcome.safetyScore;

    if (safetyScore >= AUTO_APPROVE_THRESHOLD) {
      record.tier = 'APPROVED';
      record.promotedAt = Date.now() / 1000;
      return 'APPROVED';
    } else if (safetyScore >= QUARANTINE_THRESHOLD) {
      record.tier = 'QUARANTINED';
      record.quarantineReason = `Safety score ${safetyScore.toFixed(2)} below auto-approve threshold`;
      return 'QUARANTINED';
    } else {
      // Very low safety — remove from pipeline
      this._records = this._records.filter((r) => r.recordId !== recordId);
      return 'REJECTED';
    }
  }

  /**
   * Manually approve a quarantined record (human review).
   */
  approveRecord(recordId: string, reviewNotes: string = ''): boolean {
    const record = this._records.find((r) => r.recordId === recordId);
    if (!record || record.tier !== 'QUARANTINED') return false;

    record.tier = 'APPROVED';
    record.promotedAt = Date.now() / 1000;
    record.reviewNotes = reviewNotes;
    return true;
  }

  /**
   * Reject a quarantined record (human review).
   */
  rejectRecord(recordId: string, reason: string = ''): boolean {
    const record = this._records.find((r) => r.recordId === recordId);
    if (!record || record.tier !== 'QUARANTINED') return false;

    this._records = this._records.filter((r) => r.recordId !== recordId);
    return true;
  }

  // -------------------------------------------------------------------------
  //  Tier 3: Generate SFT Pairs — APPROVED → SFT
  // -------------------------------------------------------------------------

  /**
   * Generate SFT pairs from all approved records that haven't been processed.
   * Returns the number of new pairs generated.
   */
  generateSFTPairs(): number {
    const approved = this._records.filter((r) => r.tier === 'APPROVED');
    const existingSourceIds = new Set(this._pairs.map((p) => p.sourceRecordId));

    let count = 0;
    for (const record of approved) {
      if (existingSourceIds.has(record.recordId)) continue;

      const pair = this._eventToSFTPair(record);
      if (pair && pair.qualityScore >= MIN_QUALITY_SCORE) {
        this._pairs.push(pair);
        count++;
      }
    }

    return count;
  }

  /**
   * Convert a dataset record into an SFT instruction/response pair.
   */
  private _eventToSFTPair(record: DatasetRecord): SFTPair | null {
    const event = record.event;
    const ctx = event.context;
    const action = event.action;
    const outcome = event.outcome;

    // Build the instruction (context prompt)
    const instruction = [
      `[${event.eventType.toUpperCase()}]`,
      `Location: ${ctx.location}`,
      `Companion HP: ${(ctx.companionHpRatio * 100).toFixed(0)}%`,
      `Bond Level: ${ctx.bondLevel}`,
      `Evolution: ${ctx.evolutionStage}`,
      `Formation: ${ctx.formationType}`,
      `Enemies: ${ctx.enemyCount}`,
      ctx.statusEffects.length > 0
        ? `Status: ${ctx.statusEffects.join(', ')}`
        : null,
      `Tongue: [${ctx.tongueVector.map((v) => v.toFixed(2)).join(', ')}]`,
      ``,
      `What action should the companion take?`,
    ]
      .filter(Boolean)
      .join('\n');

    // Build the response (what the companion did)
    const response = [
      `Action: ${action.description}`,
      `Result: ${outcome.description}`,
      outcome.success ? 'Outcome: SUCCESS' : 'Outcome: FAILURE',
      `Effect: ${outcome.numericResult > 0 ? '+' : ''}${outcome.numericResult}`,
    ].join('\n');

    // Quality score = safety × success_factor × confidence
    const successFactor = outcome.success ? 1.0 : 0.3;
    const qualityScore =
      outcome.safetyScore * successFactor * Math.max(action.confidence, 0.5);

    return {
      pairId: `sft_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
      sourceRecordId: record.recordId,
      companionId: this._companionId,
      instruction,
      response,
      category: event.eventType,
      qualityScore: Math.min(1.0, qualityScore),
      timestamp: Date.now() / 1000,
    };
  }

  // -------------------------------------------------------------------------
  //  HuggingFace Push
  // -------------------------------------------------------------------------

  /**
   * Create a batch of unpushed SFT pairs for HuggingFace upload.
   * Returns null if no pairs are ready.
   */
  createBatch(): HFBatch | null {
    const unpushed = this._pairs.filter(
      (p) => !this._batches.some((b) => b.pairs.some((bp) => bp.pairId === p.pairId))
    );

    if (unpushed.length === 0) return null;

    const batch: HFBatch = {
      batchId: `batch_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
      companionId: this._companionId,
      pairs: [...unpushed],
      datasetRepo: this._datasetRepo,
      pushedAt: null,
      status: 'pending',
    };

    this._batches.push(batch);
    return batch;
  }

  /**
   * Mark a batch as pushed (called after successful HF upload).
   */
  markBatchPushed(batchId: string): boolean {
    const batch = this._batches.find((b) => b.batchId === batchId);
    if (!batch || batch.status !== 'pending') return false;

    (batch as { status: string }).status = 'pushed';
    (batch as { pushedAt: number | null }).pushedAt = Date.now() / 1000;
    return true;
  }

  /**
   * Format batch as JSONL lines for HuggingFace upload.
   */
  batchToJSONL(batchId: string): string | null {
    const batch = this._batches.find((b) => b.batchId === batchId);
    if (!batch) return null;

    return batch.pairs
      .map((pair) =>
        JSON.stringify({
          id: pair.pairId,
          category: pair.category,
          instruction: pair.instruction,
          response: pair.response,
          metadata: {
            source: 'spiral_forge_gameplay',
            version: '1.0.0',
            companion_id: pair.companionId,
            quality_score: pair.qualityScore,
            timestamp: pair.timestamp,
          },
        })
      )
      .join('\n');
  }

  // -------------------------------------------------------------------------
  //  Fine-Tune Job Management
  // -------------------------------------------------------------------------

  /**
   * Check if there are enough approved pairs to trigger a fine-tune.
   */
  canTriggerFineTune(): boolean {
    const pushedPairs = this._batches
      .filter((b) => b.status === 'pushed')
      .reduce((sum, b) => sum + b.pairs.length, 0);
    return pushedPairs >= FINE_TUNE_THRESHOLD;
  }

  /**
   * Create a fine-tune job. Returns the job, or null if not enough data.
   */
  createFineTuneJob(): FineTuneJob | null {
    if (!this.canTriggerFineTune()) return null;

    const pushedBatches = this._batches.filter((b) => b.status === 'pushed');
    const totalPairs = pushedBatches.reduce((sum, b) => sum + b.pairs.length, 0);

    const job: FineTuneJob = {
      jobId: `ft_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
      companionId: this._companionId,
      datasetRepo: this._datasetRepo,
      modelRepo: this._modelRepo,
      batchIds: pushedBatches.map((b) => b.batchId),
      totalPairs,
      startedAt: Date.now() / 1000,
      completedAt: null,
      status: 'queued',
      checkpointHash: null,
    };

    this._jobs.push(job);
    return job;
  }

  /**
   * Mark a fine-tune job as completed with a model checkpoint.
   */
  completeFineTuneJob(jobId: string, checkpointHash: string): boolean {
    const job = this._jobs.find((j) => j.jobId === jobId);
    if (!job || job.status === 'completed') return false;

    job.status = 'completed';
    job.completedAt = Date.now() / 1000;
    job.checkpointHash = checkpointHash;
    return true;
  }

  // -------------------------------------------------------------------------
  //  Queries
  // -------------------------------------------------------------------------

  /** Get all records in a specific tier */
  getRecordsByTier(tier: DatasetTier): readonly DatasetRecord[] {
    return this._records.filter((r) => r.tier === tier);
  }

  /** Get all SFT pairs */
  getPairs(): readonly SFTPair[] {
    return this._pairs;
  }

  /** Get all batches */
  getBatches(): readonly HFBatch[] {
    return this._batches;
  }

  /** Get all fine-tune jobs */
  getJobs(): readonly FineTuneJob[] {
    return this._jobs;
  }

  /** Get total record count */
  get totalRecords(): number {
    return this._records.length;
  }

  /** Get total pair count */
  get totalPairs(): number {
    return this._pairs.length;
  }

  /** Summary for UI display */
  summary(): {
    companionId: string;
    rawCount: number;
    quarantinedCount: number;
    approvedCount: number;
    sftPairCount: number;
    batchesPushed: number;
    batchesPending: number;
    fineTuneJobs: number;
    fineTuneCompleted: number;
    latestCheckpoint: string | null;
    datasetRepo: string;
    modelRepo: string;
  } {
    const latestCompletedJob = this._jobs
      .filter((j) => j.status === 'completed')
      .sort((a, b) => (b.completedAt ?? 0) - (a.completedAt ?? 0))[0];

    return {
      companionId: this._companionId,
      rawCount: this._records.filter((r) => r.tier === 'RAW').length,
      quarantinedCount: this._records.filter((r) => r.tier === 'QUARANTINED').length,
      approvedCount: this._records.filter((r) => r.tier === 'APPROVED').length,
      sftPairCount: this._pairs.length,
      batchesPushed: this._batches.filter((b) => b.status === 'pushed').length,
      batchesPending: this._batches.filter((b) => b.status === 'pending').length,
      fineTuneJobs: this._jobs.length,
      fineTuneCompleted: this._jobs.filter((j) => j.status === 'completed').length,
      latestCheckpoint: latestCompletedJob?.checkpointHash ?? null,
      datasetRepo: this._datasetRepo,
      modelRepo: this._modelRepo,
    };
  }
}

// ---------------------------------------------------------------------------
//  Adventure Session — Ties energy tokens + events + training together
// ---------------------------------------------------------------------------

/**
 * An adventure session links Energy Token consumption to training data generation.
 * When a player pays tokens for a dungeon run, this session captures all events,
 * pipes them through the training pipeline, and tracks the full audit trail.
 */
export interface AdventureSession {
  readonly sessionId: string;
  readonly playerId: string;
  readonly companionId: string;
  readonly activityType: string;

  /** Energy token consumption record (proof of payment) */
  readonly consumptionRecord: ConsumptionRecord;

  /** Events captured during this adventure */
  readonly events: GameEvent[];

  /** Training pipeline results */
  readonly trainingResult: {
    approved: number;
    quarantined: number;
    rejected: number;
    sftPairsGenerated: number;
  } | null;

  readonly startedAt: number;
  completedAt: number | null;
}

/**
 * Create a helper to build game events for testing / production.
 */
export function createGameEvent(
  sessionId: string,
  playerId: string,
  companionId: string,
  eventType: GameEventType,
  context: EventContext,
  action: EventAction,
  outcome: EventOutcome
): GameEvent {
  return {
    eventId: `evt_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
    sessionId,
    playerId,
    companionId,
    eventType,
    timestamp: Date.now() / 1000,
    context,
    action,
    outcome,
  };
}

File: src/symphonic_cipher/scbe_aethermoore/game/training_pipeline.py
"""
Training Pipeline — Game Events → SFT → HuggingFace Fine-Tune (Python reference).

Mirrors src/game/trainingPipeline.ts.

Flow:
  1. Player PAYS Energy Tokens for an adventure.
  2. Game events stream into a 3-tier dataset: RAW → QUARANTINED → APPROVED.
  3. Approved events become SFT (Supervised Fine-Tune) pairs.
  4. SFT pairs batch and push to HuggingFace dataset repo.
  5. Fine-tune job trains the companion model.
  6. Updated model deploys back into the companion.

A3: Causality — events are time-ordered, append-only.
A5: Composition — full audit trail from payment → events → SFT → model.
"""

from __future__ import annotations

import json
import time
import uuid
from dataclasses import dataclass, field
from typing import Dict, List, Literal, Optional, Tuple

# ---------------------------------------------------------------------------
#  Event Types
# ---------------------------------------------------------------------------

GameEventType = Literal[
    "combat_action", "companion_command", "companion_response",
    "evolution_choice", "formation_change", "codex_query",
    "npc_dialogue", "item_use", "exploration_action", "tower_strategy",
]

DatasetTier = Literal["RAW", "QUARANTINED", "APPROVED"]


# ---------------------------------------------------------------------------
#  Event data structures
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class EventContext:
    tongue_vector: Tuple[float, ...]
    location: str
    companion_hp_ratio: float
    enemy_count: int
    formation_type: str
    bond_level: int
    evolution_stage: str
    status_effects: Tuple[str, ...] = ()


@dataclass(frozen=True)
class EventAction:
    action_id: str
    description: str
    source: Literal["player", "companion", "system"]
    confidence: float = 1.0


@dataclass(frozen=True)
class EventOutcome:
    success: bool
    numeric_result: float
    description: str
    tongue_shift: Optional[Tuple[float, ...]] = None
    safety_score: float = 1.0


@dataclass(frozen=True)
class GameEvent:
    event_id: str
    session_id: str
    player_id: str
    companion_id: str
    event_type: GameEventType
    timestamp: float
    context: EventContext
    action: EventAction
    outcome: EventOutcome


# ---------------------------------------------------------------------------
#  Dataset Record
# ---------------------------------------------------------------------------


@dataclass
class DatasetRecord:
    record_id: str
    event: GameEvent
    tier: DatasetTier
    created_at: float
    promoted_at: Optional[float] = None
    quarantine_reason: Optional[str] = None
    review_notes: Optional[str] = None


# ---------------------------------------------------------------------------
#  SFT Pair
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class SFTPair:
    pair_id: str
    source_record_id: str
    companion_id: str
    instruction: str
    response: str
    category: str
    quality_score: float
    timestamp: float


# ---------------------------------------------------------------------------
#  HF Batch + Fine-Tune Job
# ---------------------------------------------------------------------------


@dataclass
class HFBatch:
    batch_id: str
    companion_id: str
    pairs: List[SFTPair]
    dataset_repo: str
    pushed_at: Optional[float] = None
    status: Literal["pending", "pushed", "failed"] = "pending"


@dataclass
class FineTuneJob:
    job_id: str
    companion_id: str
    dataset_repo: str
    model_repo: str
    batch_ids: List[str]
    total_pairs: int
    started_at: float
    completed_at: Optional[float] = None
    status: Literal["queued", "running", "completed", "failed"] = "queued"
    checkpoint_hash: Optional[str] = None


# ---------------------------------------------------------------------------
#  Safety Thresholds
# ---------------------------------------------------------------------------

AUTO_APPROVE_THRESHOLD = 0.8
QUARANTINE_THRESHOLD = 0.4
MIN_QUALITY_SCORE = 0.5
FINE_TUNE_THRESHOLD = 100


# ---------------------------------------------------------------------------
#  Training Pipeline
# ---------------------------------------------------------------------------


class TrainingPipeline:
    """Manages the full game-event → SFT → HuggingFace pipeline for one companion."""

    def __init__(
        self,
        companion_id: str,
        player_id: str,
        dataset_repo: str = "",
        model_repo: str = "",
    ) -> None:
        self._companion_id = companion_id
        self._player_id = player_id
        self._dataset_repo = dataset_repo or f"scbe-companion-{companion_id}"
        self._model_repo = model_repo or f"scbe-model-{companion_id}"
        self._records: List[DatasetRecord] = []
        self._pairs: List[SFTPair] = []
        self._batches: List[HFBatch] = []
        self._jobs: List[FineTuneJob] = []

    @property
    def companion_id(self) -> str:
        return self._companion_id

    @property
    def dataset_repo(self) -> str:
        return self._dataset_repo

    @property
    def model_repo(self) -> str:
        return self._model_repo

    @property
    def total_records(self) -> int:
        return len(self._records)

    @property
    def total_pairs(self) -> int:
        return len(self._pairs)

    # ----- Tier 1: Ingest -----

    def ingest_event(self, event: GameEvent) -> str:
        record_id = f"rec_{int(time.time())}_{uuid.uuid4().hex[:6]}"
        record = DatasetRecord(
            record_id=record_id,
            event=event,
            tier="RAW",
            created_at=time.time(),
        )
        self._records.append(record)
        return record_id

    def ingest_session(
        self, events: List[GameEvent]
    ) -> Dict[str, int]:
        approved = quarantined = rejected = 0
        for event in events:
            record_id = self.ingest_event(event)
            result = self.classify_record(record_id)
            if result == "APPROVED":
                approved += 1
            elif result == "QUARANTINED":
                quarantined += 1
            else:
                rejected += 1
        return {"approved": approved, "quarantined": quarantined, "rejected": rejected}

    # ----- Tier 2: Classify -----

    def classify_record(self, record_id: str) -> str:
        record = next((r for r in self._records if r.record_id == record_id), None)
        if record is None or record.tier != "RAW":
            return "REJECTED"

        safety_score = record.event.outcome.safety_score

        if safety_score >= AUTO_APPROVE_THRESHOLD:
            record.tier = "APPROVED"
            record.promoted_at = time.time()
            return "APPROVED"
        elif safety_score >= QUARANTINE_THRESHOLD:
            record.tier = "QUARANTINED"
            record.quarantine_reason = (
                f"Safety score {safety_score:.2f} below auto-approve threshold"
            )
            return "QUARANTINED"
        else:
            self._records = [r for r in self._records if r.record_id != record_id]
            return "REJECTED"

    def approve_record(self, record_id: str, review_notes: str = "") -> bool:
        record = next((r for r in self._records if r.record_id == record_id), None)
        if record is None or record.tier != "QUARANTINED":
            return False
        record.tier = "APPROVED"
        record.promoted_at = time.time()
        record.review_notes = review_notes
        return True

    def reject_record(self, record_id: str, reason: str = "") -> bool:
        record = next((r for r in self._records if r.record_id == record_id), None)
        if record is None or record.tier != "QUARANTINED":
            return False
        self._records = [r for r in self._records if r.record_id != record_id]
        return True

    # ----- Tier 3: SFT Generation -----

    def generate_sft_pairs(self) -> int:
        approved = [r for r in self._records if r.tier == "APPROVED"]
        existing_ids = {p.source_record_id for p in self._pairs}
        count = 0
        for record in approved:
            if record.record_id in existing_ids:
                continue
            pair = self._event_to_sft_pair(record)
            if pair is not None and pair.quality_score >= MIN_QUALITY_SCORE:
                self._pairs.append(pair)
                count += 1
        return count

    def _event_to_sft_pair(self, record: DatasetRecord) -> Optional[SFTPair]:
        event = record.event
        ctx = event.context
        action = event.action
        outcome = event.outcome

        parts = [
            f"[{event.event_type.upper()}]",
            f"Location: {ctx.location}",
            f"Companion HP: {ctx.companion_hp_ratio * 100:.0f}%",
            f"Bond Level: {ctx.bond_level}",
            f"Evolution: {ctx.evolution_stage}",
            f"Formation: {ctx.formation_type}",
            f"Enemies: {ctx.enemy_count}",
        ]
        if ctx.status_effects:
            parts.append(f"Status: {', '.join(ctx.status_effects)}")
        tv = ", ".join(f"{v:.2f}" for v in ctx.tongue_vector)
        parts.append(f"Tongue: [{tv}]")
        parts.append("")
        parts.append("What action should the companion take?")
        instruction = "\n".join(parts)

        sign = "+" if outcome.numeric_result > 0 else ""
        response = "\n".join([
            f"Action: {action.description}",
            f"Result: {outcome.description}",
            "Outcome: SUCCESS" if outcome.success else "Outcome: FAILURE",
            f"Effect: {sign}{outcome.numeric_result}",
        ])

        success_factor = 1.0 if outcome.success else 0.3
        quality_score = min(
            1.0,
            outcome.safety_score * success_factor * max(action.confidence, 0.5),
        )

        return SFTPair(
            pair_id=f"sft_{int(time.time())}_{uuid.uuid4().hex[:6]}",
            source_record_id=record.record_id,
            companion_id=self._companion_id,
            instruction=instruction,
            response=response,
            category=event.event_type,
            quality_score=quality_score,
            timestamp=time.time(),
        )

    # ----- HuggingFace Push -----

    def create_batch(self) -> Optional[HFBatch]:
        batched_ids = set()
        for b in self._batches:
            for p in b.pairs:
                batched_ids.add(p.pair_id)

        unpushed = [p for p in self._pairs if p.pair_id not in batched_ids]
        if not unpushed:
            return None

        batch = HFBatch(
            batch_id=f"batch_{int(time.time())}_{uuid.uuid4().hex[:6]}",
            companion_id=self._companion_id,
            pairs=list(unpushed),
            dataset_repo=self._dataset_repo,
        )
        self._batches.append(batch)
        return batch

    def mark_batch_pushed(self, batch_id: str) -> bool:
        batch = next((b for b in self._batches if b.batch_id == batch_id), None)
        if batch is None or batch.status != "pending":
            return False
        batch.status = "pushed"
        batch.pushed_at = time.time()
        return True

    def batch_to_jsonl(self, batch_id: str) -> Optional[str]:
        batch = next((b for b in self._batches if b.batch_id == batch_id), None)
        if batch is None:
            return None
        lines = []
        for pair in batch.pairs:
            lines.append(json.dumps({
                "id": pair.pair_id,
                "category": pair.category,
                "instruction": pair.instruction,
                "response": pair.response,
                "metadata": {
                    "source": "spiral_forge_gameplay",
                    "version": "1.0.0",
                    "companion_id": pair.companion_id,
                    "quality_score": pair.quality_score,
                    "timestamp": pair.timestamp,
                },
            }))
        return "\n".join(lines)

    # ----- Fine-Tune Jobs -----

    def can_trigger_fine_tune(self) -> bool:
        pushed_pairs = sum(
            len(b.pairs) for b in self._batches if b.status == "pushed"
        )
        return pushed_pairs >= FINE_TUNE_THRESHOLD

    def create_fine_tune_job(self) -> Optional[FineTuneJob]:
        if not self.can_trigger_fine_tune():
            return None
        pushed = [b for b in self._batches if b.status == "pushed"]
        total = sum(len(b.pairs) for b in pushed)
        job = FineTuneJob(
            job_id=f"ft_{int(time.time())}_{uuid.uuid4().hex[:6]}",
            companion_id=self._companion_id,
            dataset_repo=self._dataset_repo,
            model_repo=self._model_repo,
            batch_ids=[b.batch_id for b in pushed],
            total_pairs=total,
            started_at=time.time(),
        )
        self._jobs.append(job)
        return job

    def complete_fine_tune_job(self, job_id: str, checkpoint_hash: str) -> bool:
        job = next((j for j in self._jobs if j.job_id == job_id), None)
        if job is None or job.status == "completed":
            return False
        job.status = "completed"
        job.completed_at = time.time()
        job.checkpoint_hash = checkpoint_hash
        return True

    # ----- Queries -----

    def get_records_by_tier(self, tier: DatasetTier) -> List[DatasetRecord]:
        return [r for r in self._records if r.tier == tier]

    def get_pairs(self) -> List[SFTPair]:
        return list(self._pairs)

    def get_batches(self) -> List[HFBatch]:
        return list(self._batches)

    def get_jobs(self) -> List[FineTuneJob]:
        return list(self._jobs)

    def summary(self) -> Dict:
        completed = [j for j in self._jobs if j.status == "completed"]
        completed.sort(key=lambda j: j.completed_at or 0, reverse=True)
        return {
            "companion_id": self._companion_id,
            "raw_count": sum(1 for r in self._records if r.tier == "RAW"),
            "quarantined_count": sum(1 for r in self._records if r.tier == "QUARANTINED"),
            "approved_count": sum(1 for r in self._records if r.tier == "APPROVED"),
            "sft_pair_count": len(self._pairs),
            "batches_pushed": sum(1 for b in self._batches if b.status == "pushed"),
            "batches_pending": sum(1 for b in self._batches if b.status == "pending"),
            "fine_tune_jobs": len(self._jobs),
            "fine_tune_completed": len(completed),
            "latest_checkpoint": completed[0].checkpoint_hash if completed else None,
            "dataset_repo": self._dataset_repo,
            "model_repo": self._model_repo,
        }


# ---------------------------------------------------------------------------
#  Helper: create game event
# ---------------------------------------------------------------------------


def create_game_event(
    session_id: str,
    player_id: str,
    companion_id: str,
    event_type: GameEventType,
    context: EventContext,
    action: EventAction,
    outcome: EventOutcome,
) -> GameEvent:
    return GameEvent(
        event_id=f"evt_{int(time.time())}_{uuid.uuid4().hex[:6]}",
        session_id=session_id,
        player_id=player_id,
        companion_id=companion_id,
        event_type=event_type,
        timestamp=time.time(),
        context=context,
        action=action,
        outcome=outcome,
    )
